enhancements/improve-kubeletconfig-default-configuration-check-enhancement.md

```diff
@@ -0,0 +1,278 @@
+    "imagefs.available": "15%",
+    "memory.available": "100Mi",
+    "nodefs.available": "10%",
+    "nodefs.inodesFree": "5%"
+  },
+  "evictionPressureTransitionPeriod": "5m0s",
+  "enableControllerAttachDetach": true,
+  "makeIPTablesUtilChains": true,
+  "iptablesMasqueradeBit": 14,
+  "iptablesDropBit": 15,
+  "featureGates": {
+    "APIPriorityAndFairness": true,
+    "CSIMigrationAWS": false,
+    "CSIMigrationAzureFile": false,
+    "CSIMigrationGCE": false,
+    "CSIMigrationvSphere": false,
+    "DownwardAPIHugePages": true,
+    "PodSecurity": true,
+    "RotateKubeletServerCertificate": true
+  },
+  "failSwapOn": true,
+  "memorySwap": {},
+  "containerLogMaxSize": "50Mi",
+  "containerLogMaxFiles": 5,
+  "configMapAndSecretChangeDetectionStrategy": "Watch",
+  "systemReserved": {
+    "cpu": "500m",
+    "memory": "1Gi"
+  },
+  "enforceNodeAllocatable": [
+    "pods"
+  ],
+  "volumePluginDir": "/etc/kubernetes/kubelet-plugins/volume/exec",
+  "logging": {
+    "format": "text",
+    "flushFrequency": 5000000000,
+    "verbosity": 2,
+    "options": {
+      "json": {
+        "infoBufferSize": "0"
+      }
+    }
+  },
+  "enableSystemLogHandler": true,
+  "shutdownGracePeriod": "0s",
+  "shutdownGracePeriodCriticalPods": "0s",
+  "enableProfilingHandler": true,
+  "enableDebugFlagsHandler": true,
+  "seccompDefault": false,
+  "registerNode": true,
+  "kind": "KubeletConfiguration",
+  "apiVersion": "kubelet.config.k8s.io/v1beta1"
+}
+```
+
+When we get `KubeletConfig` for each node, we will compare them, and
+only save the consistent result to a file at `/api/v1/nodes/proxy/configz`.
+
+In the CaC content, we just need to check one location `/api/v1/nodes/proxy/configz`
+for `KubeletConfig` default configuration.
+
+### User Stories
+
+* As an Site Reliability Engineer managing multiple clusters with many nodes, I'd like to run
+  compliance scans on my fleet in such a way that we have less compliance rule failures when
+  no reboot remediation has been applied.
+
+### API Extensions
+
+None is changed
+
+
+### Risks and Mitigations
+
+There are no risks or mitigations for this change.
+
+## Design Details
+
+### Open Questions [optional]
+
+
+1. Is `/api/v1/nodes/{NODE_NAME}/proxy/configz` a good place to check
+   default `KubeletConfig`?
+
+   Yes, it is a good place to check default `KubeletConfig`
+
+2. If we want to check the KubeletConfig through Kubernetes API, the rule
+   needs to change from node-rule to platform rule. Since platform rules
+   does not care about node type, it will require all nodes to have the 
+   same `KubeletConfig` configuration, wondering if that is a good idea?
+
+   Only common `KubeletConfig` configuration will be saved to 
+   `/api/v1/nodes/proxy/configz`
+
+### Test Plan
+
+It will be tested and reflected in our pre-existing CI, some `KubeletConfig` 
+checks will pass when the default `KubeletConfig` configuration is checked
+by Compliance Operator and meets the compliance requirements.
```

Comment:
Does it mean that some of our rules will be deprecated in favor of a new one?

Reply:
yes, the old kubelet rule will get deprecated, I wonder if we should exclude them from the profile instead of deleting them, so if user have situations where their KubeletConfig is different across the different pool.

Reply:
so the thing I was wondering about was tailored configs that use the old rules. Would we break them if we remove a rule?
On one hand, removing a rule is something we should support, we can claim backwards compatibility for content, that would tie is too much.
Do you know if the tailored profiles that exclude rules would keep working if the rule exists in a profile but is not selected? Maybe we could then deprecate the old rules, mark them as deprecated in release notes and then after one or two releases remove them.

(For any other rule I would not care much, but I know that the kubelet rules already cause some pain for customers so maybe we should treat them a little special)

Reply:
Is it possible to have the compliance operator work with both versions of the rule?

Reply:
I will give it a try, and see if it is okay to exclude not-selected rules

Reply:
Yes, the tailored profile still works if the rules are not selected by any profile.

Reply:
great!